File: src/endpoints/app/patch_item.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token

async def patch_item(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.PatchItem(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set where
        where = [ f'and srl={params.srl}' ]

        # check item
        count = db.get_count(
            table_name = Table.APP.value,
            where = where,
        )
        if count == 0: raise Exception('Item not found.', 204)

        # set values
        values = {}
        if params.code:
            values['code'] = params.code
        if params.name:
            values['name'] = params.name
        if params.description:
            values['description'] = params.description

        # check values
        if not bool(values): raise Exception('No values to update.', 400)

        # set sets
        placeholders = []
        if 'code' in values:
            placeholders.append('code = :code')
        if 'name' in values:
            placeholders.append('name = :name')
        if 'description' in values:
            placeholders.append('description = :description')

        # check exist code
        if 'code' in values and values['code']:
            count = db.get_count(
                table_name = Table.APP.value,
                where = [ f'code LIKE :code' ],
                values = { 'code': values['code'] },
            )
            if count > 0: raise Exception('"code" already exists.')

        # update item
        db.update_item(
            table_name = Table.APP.value,
            where = where,
            placeholders = placeholders,
            values = values,
        )

        # set result
        result = output.success({
            'message': 'Complete update app.',
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/nest/patch-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import * as messages from '@/libs/messages'
import { parseJSON, filteringObject } from '@/libs/objects'
import * as appHelper from '@/routes/app/__helper'
import * as helper from './__helper'
import type { NestModel } from './__model'

type PatchItemParams = {
  srl: number
  body: NestModel['patchItemBody']
}

export default async function patchItem({ srl, body }: PatchItemParams)
{
  try
  {
    // set assets
    const _table = `${DB.TABLE.NEST} AS n`

    // get data
    const item = db.getData({
      table: _table,
      where: `srl = ${srl}`,
    })
    if (!item.data) throw new ServiceError('No data.', { status: 204 })

    // check app_srl
    if (body.app)
    {
      const _count = appHelper.count({
        where: `srl = ${body.app}`,
      })
      if (_count <= 0)
      {
        throw new ServiceError('Invalid app.', { status: 400 })
      }
    }

    // set ready update
    let _ready: ZZ = {
      app_srl: undefined,
      code: undefined,
      name: undefined,
      description: undefined,
      json: undefined,
    }

    // check exist code
    if (body.app)
    {
      _ready.app_srl = body.app
    }
    if (body.code && item.data.code !== body.code)
    {
      _ready.code = body.code
      const _count = helper.count({
        where: `code LIKE \'${body.code}\'`,
      })
      if (_count > 0)
      {
        throw new ServiceError('"code" already exists.', { status: 400 })
      }
    }
    if (body.name)
    {
      _ready.name = body.name
    }
    if (body.description)
    {
      _ready.description = body.description
    }
    if (body.json)
    {
      const _json = parseJSON(body.json)
      if (!_json) throw new ServiceError('Invalid JSON data.', { status: 400 })
      _ready.json = JSON.stringify(_json)
    }

    // check update data
    _ready = filteringObject(_ready)
    if (Object.keys(_ready).length <= 0)
    {
      throw new ServiceError(messages.ERR_CANT_UPDATE, { status: 400 })
    }

    // update data
    db.editData({
      table: DB.TABLE.NEST,
      where: `srl = ${srl}`,
      set: [
        _ready.app_srl !== undefined && 'app_srl = $app_srl',
        _ready.code !== undefined && 'code = $code',
        _ready.name !== undefined && 'name = $name',
        _ready.description !== undefined && 'description = $description',
        _ready.json !== undefined && 'json = $json',
      ],
      values: {
        $app_srl: _ready.app_srl,
        $code: _ready.code,
        $name: _ready.name,
        $description: _ready.description,
        $json: _ready.json,
      },
    })
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to edit Nest.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
